doc/developer/design/20231113_optimizer_notice_catalog.md

```diff
@@ -0,0 +1,555 @@
+
+Conceptually, the `mz_optimizer_notices` view is defined as:
+
+```sql
+A EXCEPT ALL (B UNION C)
+```
+
+where
+
+- `A` corresponds to [all `mz_all_optimizer_notices`
+  entries](#mz_all_optimizer_notices).
+- `B` corresponds to [all `mz_all_optimizer_notices`
+  entries](#mz_all_optimizer_notices) that have a dependency that is not in
+  `mz_objects`. This ensures that whenever a dependency is deleted we also
+  implicitly delete any notices that depend on it.
+- `C` corresponds to [all `mz_all_optimizer_notices`
+  entries](#mz_all_optimizer_notices) that that have a `fingerprint` contained
+  [in `mz_ignored_optimizer_notices`](#mz_ignored_optimizer_notices).
+
+#### `mz_all_optimizer_notices`
+
+A new `BuiltinTable` that contains an unfiltered list of all optimizer notices
+emitted by the optimizer when planning materialized views and indexes. The
+schema of that table is identical to `mz_optimizer_notices` and its contents are
+[maintained directly by the adapter](#notice-generation-and-management).
+
+#### `mz_ignored_optimizer_notices`
+
+A new `BuiltinSource` used to keep track of ignored notices. Because of the
+`fingerprint` characteristics ignored notices will continue to be ignored even
+if we re-create the objects associated via `object_id` and `dependency_ids` as
+long as we use identical DDL statements.
+
+### Notice Generation and Management
+
+#### Notice Generation
+
+An optimizer pass produces a vector of notices as a side effect of an
+optimization run. In order to persist those notices in
+`mz_all_optimizer_notices` we need to be able to emit retractions for a
+previously inserted notice (for example when we delete the associated catalog
+item). However, the notice structure of the notices generated by the optimizer
+is not suitable for emitting retractions - for example:
+
+1. It doesn't have an optional `created_at` field.
+2. Some of the text fields need to be resolved against catalog items that might
+   be renamed or removed in the meantime.
+
+In order to protect ourselves against invalid retraction issues arising from
+these problems we make a distinction between the notice type produced by the
+optimizer (`RawOptimizerNotice`) and the notice type stored in the
+`CatalogPlans` struct (`OptimizerNotice`).
+
+Further, the type of the notices stored in `DataflowMeatinfo::optimizer_notices`
+depends on the execution stage - within the optimizer we still need to use
+`RawOptimizerNotice`, but before storing the final result in the `CatalogPlans`
+struct we need to move to a `OptimizerNotice` vector (done with a
+`CatalogState::render_notices` call). We use the `CatalogPlans` entries to
+generate insertions and deletions in the `mz_all_optimizer_notices` table.
+
+Also, since we expect to have more than two notice types in the long term, we
+reorganize the code a bit so adding a new notice is easier in the future. The
+required steps are as follows:
```

Comment:
Can you go into more detail about how we would issue retractions? For example, if a dependent object was dropped, or an object was renamed, how do we issue a retraction, even by using the Catalog? The previous dependent ID is gone from the catalog, as is the previous name.

If there's already code for this in your PR that handles this scenario, I'm more than happy to read through that :)

Reply:
There are two ways we remove remove / supress notices from the `mz_optimizer_notices` view:

1. When an entry referenced in `object_id` is deleted, we remove the notice [at the end of `Catalog::transact`](https://github.com/MaterializeInc/materialize/pull/23360/files#diff-6de498e18d68f9aa0a688e2b6bef9b0bc4057ba40284926d94126c93670ec481R1193-R1201).
2. When an entry referenced in `dependency_ids` is deleted, the [current `mz_optimizer_notices` view definition](https://github.com/MaterializeInc/materialize/pull/23360/files#diff-e31a13cbb126bb7fe119e3bf9790c20249c84db6408ea6bf548368c64428e8f6R51-R69) will filter out the enclosing notice.

In the long term I would like to emit eager retractions also for (2), but for the the sake of merging an MVP that would allow me to put together an internal console app and expose the feature to field engineering for beta testing I thought this discrepancy won't hurt too much.